### ISSUE TITLE
DirectoryAnalyzer and DirectoryWatcher should support multiprocessing (or GPU)

### ISSUE THREAD

**Issue by LimitlessGreen** [joeweiss/birdnetlib#48]
I love the idea of DirectoryAnalyzer and DirectoryWatcher. But as long as they only run on one CPU core, it's not usable for me.
I helped myself with multiprocessing, but that should only be a temporary solution as long as it isn't built directly into birdnetlib.

```python
import os
from multiprocessing import Pool
from birdnetlib import Recording
from birdnetlib.analyzer import Analyzer
from datetime import datetime


def analyze_file(file_path):
    try:
        analyzer = Analyzer()
        recording = Recording(
            analyzer,
            file_path,
            lat=35.4244,
            lon=-120.7463,
            date=datetime(year=2022, month=5, day=10),
            min_conf=0.25,
        )
        recording.analyze()
        print(f"Processed {file_path}")
        return recording
    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        return []


def analyze_directory(dir_path):
    pool = Pool()
    recordings = []
    for subdir, dirs, files in os.walk(dir_path):
        for file in files:
            if file.lower().endswith(".wav"):
                file_path = os.path.join(subdir, file)
                recordings.append(pool.apply_async(analyze_file, args=(file_path,)))
    pool.close()
    pool.join()

    return recordings


if __name__ == "__main__":
    dir_path = "./workdir/"
    recordings = analyze_directory(dir_path)
    print(f"Finished analyzing {len(recordings)} files")

```

**Comment by joeweiss** [comment]
Thanks @LimitlessGreen for opening the issue and providing your workaround. I've edited the title to represent the change that needs to be made to the project. 

I plan to add support for multiprocessing, but it will need to be optional. I know of several users that require one-process-per-analyzer as part of their larger workflow. In those cases, the workflow uses a database as the queue, and runs a separate process for each available CPU core. I believe the rationale for that requirement is to isolate out-of-memory failures from taking down the entire queue.

Adding multiprocessing to `DirectoryAnalyzer` should be the next step in this issue. Would an interface like the following work for your use case?

```python
directory = DirectoryAnalyzer(
    "/Birds/mp3_dir",
    patterns=["*.mp3", "*.wav"],
    processes=4
)
directory.on_analyze_complete = on_analyze_complete
directory.process()

```

**Comment by LimitlessGreen** [comment]
> Adding multiprocessing to `DirectoryAnalyzer` should be the next step in this issue. Would an interface like the following work for your use case?
> 
> ```python
> directory = DirectoryAnalyzer(
>     "/Birds/mp3_dir",
>     patterns=["*.mp3", "*.wav"],
>     processes=4
> )
> directory.on_analyze_complete = on_analyze_complete
> directory.process()
> ```

Yes, this interface looks good to me :octocat: :smiley:

**Comment by joeweiss** [comment]
This feature is currently parked. The issue is that data has to be pickleable when passed in and out of the Pool child processes. This isn't possible at the moment in the `DirectoryAnalyzer` class because the `process` method needs access to the Analyzer class(es) in particular. Analyzer loads and runs the actual TensorFlow models, and the initializing takes a non-trivial time on start-up. 

Currently, in `birdnetlib` analyzer start-up only happens once no matter how many audio files are eventually processed.

In the example [above](https://github.com/joeweiss/birdnetlib/issues/48#issue-1662132824), the data passed to and returned from `analyze_file` is indeed pickleable, but at the cost of having to initialize the Analyzer with every single file.

Notes to read up on:
https://towardsdatascience.com/multiprocessing-and-pickle-how-to-easily-fix-that-6f7e55dee29d

Also see: https://pypi.org/project/dill/

I'll investigate more as time allows.

As a current workaround, one option is to split up large groups of files or use a centralized queue, then run multiple instances of python with something like supervisor.

**Comment by joeweiss** [comment]
On a second look, this is possible by using a Manager.Queue for the file path listing. Each process will still have to initialize Tensorflow and the Analyzer class(es), but that is preferable to the current method.

My thoughts are to add a dedicated `DirectoryMultiProcessAnalyzer`, as I'm inclined to leave `DirectoryAnalyzer` as a single process implementation.

**Comment by joeweiss** [comment]
More notes to myself on this issue ...

The primary use-case for `DirectoryWatcher` was to provide an easy implementation for handling files as they were recorded. I'm planning to leave `DirectoryWatcher` as a single-process, since even an older Raspberry Pi 3 can keep up with realtime recording saved into a watched directory.

The implementation for `DirectoryMultiProcessAnalyzer` is in PR #52. I'll take another review of it later this week, then package it for release in 0.5.

**Comment by joeweiss** [comment]
And here's [an example](https://github.com/joeweiss/birdnetlib/blob/d2747412c7fd0249db11355c3c2bce3b5150d3e9/examples/batch_multiprocessing_directory.py) of using `DirectoryMultiProcessAnalyzer`.

**Comment by joeweiss** [comment]
`DirectoryMultiProcessAnalyzer` has been released in 0.5.0. 

I'm open to reviewing a PR for a multiprocessing implementation of `DirectoryMultiProcessWatcher` down the road. If anyone is interested in taking a stab at it, please open a new issue before submitting the pull request.
